libeos-updater-flatpak-installer/perform-flatpak-actions.c

```diff
@@ -224,7 +222,106 @@ try_uninstall_application (FlatpakInstallation  *installation,
+    flatpak_installation_list_remotes (installation, NULL, error);
+  g_autoptr(GHashTable) uninstall_dependencies_set =
+    g_hash_table_new_full (euu_flatpak_ref_hash,
+                           euu_flatpak_ref_equal,
+                           g_object_unref,
+                           NULL);
+
+  for (gsize i = 0; i < remotes->len; ++i)
+    {
+      FlatpakRemote *remote = g_ptr_array_index (remotes, i);
+      g_autoptr(GError) local_error = NULL;
+      g_autoptr(GPtrArray) installed_related_refs =
+        flatpak_installation_list_installed_related_refs_sync (installation,
+                                                               flatpak_remote_get_name (remote),
+                                                               formatted_ref,
+                                                               NULL,
+                                                               &local_error);
+
+      if (installed_related_refs == NULL)
+        {
+          /* If we hit this we are going to get the same error for all
+           * remotes, so just return now */
+          if (g_error_matches (local_error, FLATPAK_ERROR, FLATPAK_ERROR_NOT_INSTALLED))
+            return g_steal_pointer (&uninstall_dependencies_set);
+
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return NULL;
+        }
+
+      for (gsize j = 0; j < installed_related_refs->len; ++j)
+        {
+          FlatpakRelatedRef *related_ref = g_ptr_array_index (installed_related_refs, j);
+
+          if (flatpak_related_ref_should_delete (related_ref))
+            g_hash_table_insert (uninstall_dependencies_set,
+                                 g_object_ref (related_ref),
+                                 NULL);
+        }
+    }
+
+  return g_steal_pointer (&uninstall_dependencies_set);
+}
+
+static gboolean
+try_uninstall_application (FlatpakInstallation  *installation,
+                           FlatpakRef           *ref,
+                           GError              **error)
+{
+  g_autofree gchar *formatted_ref = flatpak_ref_format_ref (ref);
+  g_autoptr(GHashTable) uninstall_dependencies_set = NULL;
+  GHashTableIter iter;
+  gpointer key, value;
+
+  g_message ("Attempting to uninstall %s", formatted_ref);
+
+  uninstall_dependencies_set =
+    list_autodelete_dependencies (installation,
+                                  ref,
+                                  error);
+
+  if (uninstall_dependencies_set == NULL)
+    return FALSE;
+
+  if (!uninstall_ignoring_not_installed (installation,
+                                         ref,
+                                         error))
+    return FALSE;
+
+  g_hash_table_iter_init (&iter, uninstall_dependencies_set);
+  while (g_hash_table_iter_next (&iter, &key, &value))
+    {
+      FlatpakRef *uninstall_ref = key;
+      g_autofree gchar *formatted_related_ref = flatpak_ref_format_ref (uninstall_ref);
+
+      g_debug ("Uninstalling autodelete dependency %s of %s",
```

Comment:
So, hmm... what happens if an extension is marked as autodelete but is shared between more than one Flatpak...?

Reply:
I'm not sure. Will need to check what flatpak itself does

Reply:
If this is broken, let's split it out to a separate ticket and possibly even wait for Flatpak to have one real solution in the transaction stuff. We're not uninstalling anything with autodelete extensions for the moment, and it's not something we have to worry about getting right before the reboot - it's only run in the "updated" updater after reboot.

Reply:
So flatpak just unconditionally adds the related ref to the uninstall list: https://github.com/flatpak/flatpak/blob/70bb192204e733190d0b7d78773d5cbde385a5cb/common/flatpak-dir.c#L10777

Incidentally, `app` also uninstalls any extension ending in `Debug` and `Locale` too - hurrah for `lib` and `app` divergence :/ .

To be honest, I think his behavior is fine. From what I gather the "autodelete" property is something that exists on the extension point definition and not the extension itself. So if two apps could plausibly use the same extension, it wouldn't be marked as autodelete in one app.

Reply:
I do see your point about splitting it out into a separate ticket. That's easily done, if we're happy to go down that road.